iox_query/src/physical_optimizer/sort/push_sort_through_union.rs

```diff
@@ -0,0 +1,381 @@
+/// ```
+/// will become:
+/// ```text
+/// DeduplicateExec: [col2@1 ASC,col1@0 ASC,time@3 ASC,__chunk_order@4 ASC]
+///   SortPreservingRepartitionExec: partitioning=Hash([col2@1, col1@0, time@3, __chunk_order@4], 8), input_partitions=8
+///     SortPreservingRepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=4
+///       UnionExec
+///         SortExec: expr=[col2@1 ASC,col1@0 ASC,time@3 ASC,__chunk_order@4 ASC]
+///           RecordBatchesExec: batches_groups=2 batches=0 total_rows=0
+///         ParquetExec: file_groups={2 groups: [[1.parquet], [2.parquet]]}, projection=[col1, col2, field1, time, __chunk_order], output_ordering=[col2@1 ASC, col1@0 ASC, time@3 ASC, __chunk_order@4 ASC]
+/// ```
+///
+/// There is a tension between:
+/// - Wanting to do sorts in parallel
+/// - Sorting fewer rows
+///
+/// DataFusion will not push down a sort through a `RepartitionExec`
+/// because it could reduce the parallelism of the sort. However,
+/// in IOx, unsorted children of `UnionExec` will tend to be
+/// [`RecordBatchesExec`] which is likely to have many fewer rows than
+/// other children which will tend to be [`ParquetExec`].
+/// So making this transformation will generally have a dramatic effect
+/// on the amount of data being sorted.
+///
+/// [`RecordBatchesExec`]: crate::provider::RecordBatchesExec
+/// [`ParquetExec`]: datafusion::datasource::physical_plan::ParquetExec
+pub(crate) struct PushSortThroughUnion;
+
+impl PhysicalOptimizerRule for PushSortThroughUnion {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        plan.transform_up(&|plan| {
+            let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() else {
+                return Ok(Transformed::No(plan));
+            };
+
+            if !sort_can_be_pushed_down(sort_exec) {
+                return Ok(Transformed::No(plan));
+            }
+
+            let mut plan = Arc::clone(sort_exec.input());
+            let mut rewriter = SortRewriter {
+                ordering: sort_exec.output_ordering().unwrap().to_vec(),
+            };
+
+            plan = plan.rewrite(&mut rewriter)?;
+
+            // As a sanity check, make sure plan has the same ordering as before.
+            // If this fails, there is a bug in this optimization.
+            let required_order = sort_exec.output_ordering().map(sort_exprs_to_requirement);
+            if !ordering_satisfy_requirement(
+                plan.output_ordering(),
+                required_order.as_deref(),
+                || plan.equivalence_properties(),
+                || plan.ordering_equivalence_properties(),
+            ) {
+                return Err(datafusion::error::DataFusionError::External(
```

Comment:
This seems like it would be the result of a bug so an internal error would be warranted

```suggestion
                return Err(datafusion::error::DataFusionError::Internal(
```